File: hot_100/39. 组合总和.go
package main

// 39. 组合总和

// 给定一个无重复元素的数组candidates和一个目标数target，找出candidates中所有可以使数字和为target的组合。
// candidates中的数字可以无限制重复被选取。
//
// 说明：
// 所有数字（包括target）都是正整数。
// 解集不能包含重复的组合。

// func main() {

// 	// aa := []int{3, 5, 6, 1, 2, 8, 3, 4}
// 	// sort1(aa, 0, len(aa)-1)
// 	// fmt.Println(aa)

// 	// 回溯法可以求出左右解是什么
// 	aa := combinationSum([]int{3, 5, 6, 1, 2, 8, 3, 4}, 7)
// 	fmt.Println(len(aa))

// 	// 动态规划只能求出解的个数
// 	bb := combinationSum1([]int{3, 5, 6, 1, 2, 8, 3, 4}, 7)
// 	fmt.Println(bb)

// }

// 见Hot100 518(同类问题，动态规划)
// 1. 回溯法+剪枝 -- 可以求出每个解是什么
func combinationSum(candidates []int, target int) [][]int {
	if candidates == nil || len(candidates) == 0 {
		return nil
	}

	// 先排序数组，方便剪枝
	sort1(candidates, 0, len(candidates)-1)

	// 回溯求解
	result := make([][]int, 0)
	res := make([]int, 0)
	dfs2(candidates, 0, target, 0, &res, &result)

	return result
}

// candidates: 总数组
// currentSum: 上一层的和
// target: 目标值
// start: 从哪个位置开始当前层的数字遍历
// res: 路径
// result: 总结果
func dfs2(candidates []int, currentSum, target int, start int, res *[]int, result *[][]int) {

	// 路径结束条件 -- 若总和已经等于目标值，将路径粗存入最终结果中并返回
	if currentSum == target {
		tmp := make([]int, len(*res))
		copy(tmp, *res)
		*result = append(*result, tmp)

		return
	}

	num := 0 // 当前数字
	for i := start; i < len(candidates); i++ {

		num = candidates[i]

		// 剪枝 -- 由于数组已排序，若加上当前数字都已经比目标值大了，那么加上后面的数字更大
		if currentSum+num > target {
			break
		}

		// 将当前数字加到路径中然后
		*res = append(*res, num)
		dfs2(candidates, currentSum+num, target, i, res, result)
		*res = (*res)[:len(*res)-1]
	}
}

func sort1(arr []int, left, right int) {
	if left >= right || arr == nil || len(arr) == 0 {
		return
	}

	l, r := left, right
	tmp := arr[left]

	for l < r {
		for arr[r] >= tmp && l < r {
			r--
		}
		for arr[l] <= tmp && l < r {
			l++
		}
		if l < r {
			arr[l], arr[r] = arr[r], arr[l]
		}
	}
	arr[l], arr[left] = tmp, arr[l]

	sort1(arr, left, l-1)
	sort1(arr, l+1, right)
}

// 动态规划 -- 只是求出解的个数
// 2. 背包问题 - 组合问题-- dp[i] += dp[i-num]
func combinationSum1(candidates []int, target int) int {
	if candidates == nil || len(candidates) == 0 {
		return 0
	}

	dp := make([]int, target+1)
	dp[0] = 1

	for _, num := range candidates {
		for i := 1; i <= target; i++ {
			if i >= num {
				dp[i] += dp[i-num]
			}
		}
	}

	return dp[target]
}

File: C++/39.组合总和.cpp

// 39. 组合总和
// 给你一个 无重复元素 的整数数组 candidates 和一个目标整数 target ，
// 找出 candidates 中可以使数字和为目标数 target 的 所有 不同组合 ，
// 并以列表形式返回。你可以按 任意顺序 返回这些组合。
//
// candidates 中的 同一个 数字可以 无限制重复被选取 。
// 如果至少一个数字的被选数量不同，则两种组合是不同的。 
//
// 对于给定的输入，保证和为 target 的不同组合数少于 150 个。

#include <algorithm>
#include <string>
#include <vector>

class Solution {
public:
    // 1. 回溯法
    // 2. 动态规划 -- 完全背包问题
    std::vector<std::vector<int>> combinationSum(std::vector<int>& candidates, int target) {
        // 此题说明了数组中无重复元素, 所以不用先排序进行剪枝 -- 见40
        // quick_sort(candidates, 0, candidates.size() - 1);

        std::vector<int> res;
        std::vector<std::vector<int>> result;
        dfs(candidates, 0, 0, target, res, result);
        return result;
    }

    void dfs(std::vector<int>& candidates, int start, int current_sum, int target, std::vector<int>& res, std::vector<std::vector<int>>& result) {
        if (current_sum == target) {
            result.push_back(res);
            return;
        }

        for (int i = start; i < candidates.size(); ++i) {
            if (current_sum + candidates[i] > target) {
                // 剪枝
                return;
            }

            // 此题说明了数组中无重复元素, 所以不用跳过重复元素
            // if (i > start && candidates[i] == candidates[i - 1]) {
            //     // 跳过重复元素
            //     continue;
            // }

            res.push_back(candidates[i]);
            dfs(candidates, i, current_sum + candidates[i], target, res, result);
            res.pop_back();
        }
    }

    // 2. 动态规划 -- 组合问题 -- 完全背包问题 -- 只能求出解的个数, 无法求出每个解是什么
    int combinationSum2(std::vector<int>& candidates, int target) {
        std::vector<int> dp(target + 1, 0);  // dp[i] += dp[i - num]
        dp[0] = 0;

        for (int num : candidates) {             // 遍历物品
            for (int i = 1; i <= target; i++) {  // 遍历背包容量 -- 若每个数字只能使用1次, 则应该倒序遍历, 如40
                if (i >= num) {
                    dp[i] += dp[i - num];
                }
            }
        }

        return dp[target];
    }
};
